common/src/main/java/com/microsoft/identity/common/internal/cache/FociOAuth2TokenCache.java

```diff
@@ -0,0 +1,200 @@
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.cache;
+
+import android.content.Context;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.microsoft.identity.common.BaseAccount;
+import com.microsoft.identity.common.internal.dto.AccessTokenRecord;
+import com.microsoft.identity.common.internal.dto.AccountRecord;
+import com.microsoft.identity.common.internal.dto.Credential;
+import com.microsoft.identity.common.internal.dto.IdTokenRecord;
+import com.microsoft.identity.common.internal.dto.RefreshTokenRecord;
+import com.microsoft.identity.common.internal.logging.Logger;
+import com.microsoft.identity.common.internal.providers.oauth2.AuthorizationRequest;
+import com.microsoft.identity.common.internal.providers.oauth2.OAuth2Strategy;
+import com.microsoft.identity.common.internal.providers.oauth2.TokenResponse;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class FociOAuth2TokenCache
+        <GenericOAuth2Strategy extends OAuth2Strategy,
+                GenericAuthorizationRequest extends AuthorizationRequest,
+                GenericTokenResponse extends TokenResponse,
+                GenericAccount extends BaseAccount,
+                GenericRefreshToken extends com.microsoft.identity.common.internal.providers.oauth2.RefreshToken>
+        extends MsalOAuth2TokenCache<GenericOAuth2Strategy, GenericAuthorizationRequest, GenericTokenResponse, GenericAccount, GenericRefreshToken> {
+
+    private static final String TAG = FociOAuth2TokenCache.class.getSimpleName();
+
+    /**
+     * Constructs a new OAuth2TokenCache.
+     *
+     * @param context The Application Context of the consuming app.
+     */
+    public FociOAuth2TokenCache(final Context context,
+                                final IAccountCredentialCache accountCredentialCache,
+                                final IAccountCredentialAdapter<
+                                        GenericOAuth2Strategy,
+                                        GenericAuthorizationRequest,
+                                        GenericTokenResponse,
+                                        GenericAccount,
+                                        GenericRefreshToken> accountCredentialAdapter) {
+        super(context, accountCredentialCache, accountCredentialAdapter);
+    }
+
+    /**
+     * Loads the tokens available for the supplied client criteria.
+     *
+     * @param clientId      The current client's id.
+     * @param target        The desired scopes.
+     * @param accountRecord The current account.
+     * @return An ICacheRecord containing the account. If a matching id token is available
+     * (for the provided clientId), it is returned. If a matching access token is available
+     * (for the provided client id), it is also returned. If a matching refresh token is available
+     * it is returned.
+     */
+    public ICacheRecord loadByFamilyId(@Nullable final String clientId,
```

Comment:
Discussed offline but posting here: short answer is we don't.

The broker controller/strategy should call the `BrokerOAuth2TokenCache`'s `load()` function and inspect the result.

If the result contains an `AccountRecord`, `IdTokenRecord`, `AccessTokenRecord`, and `RefreshTokenRecord` then the result is OK to use. The caller should still check the expiry of the `AccessTokenRecord` before returning the result to the caller, refreshing as necessary...

If the result contains _only_ an `AccountRecord` then we had no tokens in the cache and the library should do some equivalent of `AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED`

If the result contains _only_ a `RefreshTokenRecord` then the caller should [attempt to refresh the access token](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow#refresh-the-access-token). If it works, call `BrokerOAuth2TokenCache#save()` with the result. If it fails, throw some equivalent of `AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED` 

@kreedula ^^ This should be the calling pattern that you and I had outlined in our previous discussion. Please chime-in with any questions, corrections, or otherwise.